### PR TITLE
Subgraph Adaptation in ALM

### PR DESCRIPTION
**Pull request by t0mcr8se**

### CODE REVIEW

File: alm/src/utils/contract.ts [hunk 1]
```diff
@@ -59,40 +68,43 @@ export const getRequiredSignatures = async (
...
+  //   const { validator } = e.returnValues
+  //   if (e.event === 'ValidatorRemoved') {
+  //     validatorList.add(validator)
+  //   } else if (e.event === 'ValidatorAdded') {
+  //     validatorList.delete(validator)
+  //   }
+  // })
+
+  // return Array.from(validatorList)
+  return ['0x6F7648D9B6203fF9e6289BBAae3B5738Ad27a0b4']
```
Comment (leonprou) [inline alm/src/utils/contract.ts]: please put this into config

File: alm/scripts/createSnapshots.js [hunk 2]
```diff
@@ -102,11 +104,19 @@ const generateSnapshot = async (side, url, bridgeAddress) => {
...
+  if (typeof str !== 'string' || str === '') return false
+  return true
+}
+
 const main = async () => {
-  await Promise.all([
-    generateSnapshot('home', COMMON_HOME_RPC_URL, COMMON_HOME_BRIDGE_ADDRESS),
-    generateSnapshot('foreign', COMMON_FOREIGN_RPC_URL, COMMON_FOREIGN_BRIDGE_ADDRESS)
-  ])
+  let promises = []
```
Comment (leonprou) [inline alm/scripts/createSnapshots.js]: this can be const

File: alm/scripts/createSnapshots.js [hunk 3]
```diff
@@ -102,11 +104,19 @@ const generateSnapshot = async (side, url, bridgeAddress) => {
...
+  // If we're using subgraphs, there is no need for snapshots. This solution was introduced because of the bsc rpc limits for eth_getLogs method
+  if (!isNonEmptyString(HOME_SUBGRAPH_URL))
```
Comment (leonprou) [inline alm/scripts/createSnapshots.js:115]: looks cool

File: alm/src/hooks/useMessageConfirmations.ts [hunk 4]
```diff
@@ -192,8 +194,10 @@ export const useMessageConfirmations = ({
         })
       }
 
-      homeBlockNumberProvider.start(home.web3)
-      const targetBlock = collectedSignaturesEvent.blockNumber + blockConfirmations
+      homeBlockNumberProvider.start(home.web3);
+      (async () => {
```
Comment (leonprou) [inline alm/src/hooks/useMessageConfirmations.ts:198]: can you explain that line?

File: alm/src/utils/contract.ts [hunk 5]
```diff
@@ -59,40 +68,43 @@ export const getRequiredSignatures = async (
   return parseInt(requiredSignatures)
 }
 
-export const getValidatorList = async (contract: Contract, blockNumber: number, snapshotProvider: SnapshotProvider) => {
+export const getValidatorList = async (
+  contract: Contract,
+  blockNumber: number,
+  fromHome: boolean,
```
Comment (leonprou) [inline alm/src/utils/contract.ts]: can you rename fromHome to `isHome`. seems more easy to follow

File: alm/src/utils/contract.ts [hunk 6]
```diff
@@ -59,40 +68,43 @@ export const getRequiredSignatures = async (
...
-    const { validator } = e.returnValues
-    if (e.event === 'ValidatorRemoved') {
-      validatorList.add(validator)
-    } else if (e.event === 'ValidatorAdded') {
-      validatorList.delete(validator)
-    }
-  })
-
-  return Array.from(validatorList)
+  // TODO: sort events
```
Comment (leonprou) [inline alm/src/utils/contract.ts]: please delete the commented code

File: alm/src/utils/getFinalizationEvent.ts [hunk 7]
```diff
@@ -31,27 +34,21 @@ export const getFinalizationEvent = async (
...
-    }
-  })
+  const events: any = await getEvents(contract, eventName, 0, fromHome, undefined, { messageId: message.id })
   if (events.length > 0) {
     const event = events[0]
-    const [txReceipt, block] = await Promise.all([
-      web3.eth.getTransactionReceipt(event.transactionHash),
-      getBlock(web3, event.blockNumber)
-    ])
+    const txReceipt = await web3.eth.getTransactionReceipt(
```
Comment (leonprou) [inline alm/src/utils/getFinalizationEvent.ts:40]: can you explain why the change here is needed

File: alm/src/config/constants.ts [hunk 8]
```diff
@@ -66,3 +69,28 @@ export const VALIDATOR_CONFIRMATION_STATUS = {
 export const SEARCHING_TX = 'Searching Transaction...'
 
 export const INCORRECT_CHAIN_ERROR = `Incorrect chain chosen. Switch to ${FOREIGN_NETWORK_NAME} in the wallet.`
+
+// Fields of the events in the subgraph
+export const SUBGRAPH_EVENT_FIELDS: { [key: string]: string[] } = {
```
Comment (leonprou) [inline alm/src/config/constants.ts:74]: liked that 👍

File: alm/.env.example [hunk 9]
```diff
@@ -1,15 +1,19 @@
-COMMON_HOME_BRIDGE_ADDRESS=0x2CA5411c4bf447Cc27CD6E6d1d046f922A27C399
-COMMON_FOREIGN_BRIDGE_ADDRESS=0x63C47c296B63bE888e9af376bd927C835014039f
+COMMON_HOME_BRIDGE_ADDRESS=0x1ee6E3E3d2DE779858728E157B3B9C488bA7b706
```
Comment (leonprou) [inline alm/.env.example]: can you leave this file untouched and open a new one like .env.bsc.example

**Review by leonprou — CHANGES_REQUESTED** [review]
Solid work 👍 
added small requests for fixes
and some questions to explain some changes you did.

File: alm/src/utils/contract.ts (hunk 1 above)
Comment (t0mcr8se) [reply alm/src/utils/contract.ts]: we don't need to store that in config, this better? https://github.com/fuseio/tokenbridge/pull/7/commits/06772c77f97c4220e5c42a0d5d88ca703b162d18

File: alm/scripts/createSnapshots.js (hunk 2 above)
Comment (t0mcr8se) [reply alm/scripts/createSnapshots.js]: https://github.com/fuseio/tokenbridge/pull/7/commits/df0c4aab116bb6b023e5df85044ed5bb41b2732a
done

File: alm/scripts/createSnapshots.js (hunk 3 above)
Comment (t0mcr8se) [reply alm/scripts/createSnapshots.js:115]: thanks, I thought it looked ugly lol

File: alm/src/hooks/useMessageConfirmations.ts (hunk 4 above)
Comment (t0mcr8se) [reply alm/src/hooks/useMessageConfirmations.ts:198]: if we're using subgraph, we do not have collectedSignaturesEvent.blockNumber because subgraph data does not include block number. so basically, we're getting the block number of the event from the tx reciept

File: alm/src/utils/contract.ts (hunk 5 above)
Comment (t0mcr8se) [reply alm/src/utils/contract.ts]: https://github.com/fuseio/tokenbridge/pull/7/commits/7ea6cb3b3ccaed3a7cb6f5f2e0b2db1268875c51
done

File: alm/src/utils/contract.ts (hunk 6 above)
Comment (t0mcr8se) [reply alm/src/utils/contract.ts]: https://github.com/fuseio/tokenbridge/commit/06772c77f97c4220e5c42a0d5d88ca703b162d18

File: alm/src/utils/getFinalizationEvent.ts (hunk 7 above)
Comment (t0mcr8se) [reply alm/src/utils/getFinalizationEvent.ts:40]: same reason as for the one with targetBlock; we can't have event.blockNumber if we're using subgraph...

File: alm/src/config/constants.ts (hunk 8 above)
Comment (t0mcr8se) [reply alm/src/config/constants.ts:74]: thanks)

File: alm/.env.example (hunk 9 above)
Comment (t0mcr8se) [reply alm/.env.example]: https://github.com/fuseio/tokenbridge/pull/7/commits/7092d5e8c1a54c80a395753c1a409dfc1e4faaf0

File: alm/scripts/createSnapshots.js (hunk 3 above)
Comment (leonprou) [reply alm/scripts/createSnapshots.js:115]: maybe, but it's nice we dont need snapshotting for thegraph

File: alm/src/hooks/useMessageConfirmations.ts (hunk 4 above)
Comment (leonprou) [reply alm/src/hooks/useMessageConfirmations.ts:198]: @masterbluecompact got it. We can add a blocknumber to the events, would that help?

**Review by leonprou — APPROVED** [review]

File: alm/src/hooks/useMessageConfirmations.ts (hunk 4 above)
Comment (t0mcr8se) [reply alm/src/hooks/useMessageConfirmations.ts:198]: yeah that would help, but it is okay the way it is right now, but could be better with block number in the subgraph data

Comment (mul53) [reply alm/src/hooks/useMessageConfirmations.ts:198]: then i will add the blocknumber to the events
